### ISSUE TITLE
Single Thread per Worker

### ISSUE THREAD

**Issue by null511** [nunit/nunit#1333]
I have a framework in development based on _Selenium_ for automated web testing that successfully runs in parallel using NUnit 3.0.1 - However this seems to be broken since 3.2.0, which includes the useful ability to output during SetUp/TearDown events.

My current setup uses `Thread.CurrentThread` against a dictionary to limit one browser instance per Worker thread, as I was unable to find a way to determine which worker created the test/fixture that is running. This seemed to be working well in 3.0.1, but is since broken in 3.2.0.

I was wondering if you think this would be a useful feature for others as well, or if you had any thoughts on implementing an option to enable this. I will most likely pull down the current source and make the modifications for my own purposes, but I could submit a PR as well if there is interest.

**Comment by oznetmaster** [comment]
In 3.2, there is now one thread per WorkItem, which in most cases corresponds to a Test method.  This is a different thread then the thread the TestFixture is running on, and only exists for the duration of the Test (including its SetUp and TearDown).

If you access your dictionary in SetUp, then it seems like it should work as desired.

**Comment by CharliePoole** [comment]
I'm not clear what you mean by "limit one browser instance per worker thread." One at a time? One for the life of the thread? Why not just require your fixture to run on it's own thread using [RequiresThread] ? That thread will then be discarded and never used again.

**Comment by null511** [comment]
Thank you for your prompt response.

@oznetmaster, I just tested this, and I'm seeing the same thread from SetUp/Test/TearDown, but a different thread per fixture, which breaks my usage. Tested by printing `Thread.CurrentThread.ManagedThreadId`

@CharliePoole, My goal is to match the _Browser_ (IWebDriver) count to the Worker count. Opening a browser instance requires a lot of overhead, so opening/closing per fixture is not an optimal solution. In 3.0.1, my tests were retrieving a browser instance from a dictionary by using `Thread.CurrentThread` as the key, creating an instance if it has not been attached.

```C#
ConcurrentDictionary<Thread, IWebDriver> instances;

[TestFixture]
void TestBase_SetUp() {
    this.Driver = instances.GetOrAdd(Thread.CurrentThread, () => new IWebDriver())
}
```

I think a more reliable solution would be to have a _WorkerContext_ available, or possibly an additional property on _TestExecutionContext_ to identify which worker the test is running under.

**Comment by null511** [comment]
I was able to implement this (using the current branch) by passing `TestWorker.Name` to a new property `WorkItem.WorkerId` before the test is executed. When it is retrieved through `TestAdapter`, the `Parent.WorkerId` will be used if the current value is null, so that tests will inherit their fixtures _WorkerId_. So far this is working great, through the _LevelOfParallelism_ attribute, users can specify the number of browser instances (ex: 2), and all tests will be queued properly to the available browser.

I have a branch ready to be pushed if this would be a useful addition to the project. I'm now using a custom NuGet package on our [my companies] internal NuGet server to distribute these additions with my framework, but it would be useful to migrate back to an official release if this could be implemented.

**Comment by CharliePoole** [comment]
Fixed by PR #1351
